### ISSUE TITLE
Option to fail builds that require polyfill

### ISSUE THREAD

**Issue by ef4** [babel/babel#2338]
Given that the polyfill is optional, it would be helpful if we could tell babel whether or not the polyfill will be available at runtime so that babel can fail builds that need it but don't have it. Otherwise it is easy to accidentally create a dependency on the polyfill that you didn't expect, and not even notice because you happen to test on a browser that has native implementations of some polyfilled features.

I'm not necessarily suggesting that babel should check for usage of polyfilled features in user code -- it is reasonable to make that the user's responsibility. I'm just suggesting that babel could refuse to introduce new dependencies on those features. For example, if the user says `for ... of`, babel would error with "`for ... of` requires a polyfill for `Symbol`, but you have enabled `--forbid-polyfill-usage`.

**Comment by sebmck** [comment]
This isn't something that's going to be added to core, sorry. What requires the polyfill is completely arbitrary so when you say "what requires the polyfill" it completely depends on the environment you're targeting. However, you can look into writing your own [babel plugin](http://babeljs.io/docs/usage/plugins) or find some way of disabling certain functions in your linter.

**Comment by ef4** [comment]
Then let me rephrase my problem statement: only babel core actually knows which browser features it is going to use to transpile a given input, so only babel core actually knows what environments are supported by the the final output. (Short of saying "have a perfect test suite", which is a trivial-but-unhelpful solution to every problem.)

Because there is no public API for asking babel "what browser features are you using?", all the internal dependencies accidentally become public API, limiting your future ability to change implementations without breaking apps.

> it completely depends on the environment you're targeting.

Yes, that is my entire point. We need public API that talks about what environments we're targeting. I can see that "needs polyfill" is too coarse grained. But "needs `Symbol`" is not, and right now babel core is the only code that actually knows which constructs transpile to code that needs `Symbol`.

**Comment by sebmck** [comment]
You want Babel to do linting. It wont be completely out of scope in the future but for now it is.

**Comment by ef4** [comment]
No, I definitely don't want linting. 

Linting post-transpiled code is not helpful to users (because they get error messages about code they didn't write, and have a hard time reversing that back to what they actually did), and linting pre-transpiled code wouldn't detect the problems I'm talking about. 

But the biggest problem with treating this as a linting problem is that it closes the door on solutions that degrade & upgrade automatically based on the targeted environment, which is going to be a bigger and bigger deal as browsers gain native support for many features that require transpilation today.

I think you're suggesting linting because it's the only solution that is easy to implement without extending the API between babel and its many transforms. I'm suggesting that expanding that API is inevitable as the feature set requiring transpilation continues to evolve. The sooner the ecosystem gains an API for plugins to declare what browser features they need and to ask what features they're allowed to use, the easier it will be to move the whole ecosystem forward together.

If you look at the history of any mature compiler, you'll see a similar progression. The target environments evolve over time, so the compilers necessarily start to grow more comprehensive support for tuning what environmental features the compiled output will depend on.

I won't belabor this thread further -- my goal is not to argue you into implementing my pet feature. My goal is to point out the inevitable problems that the future will bring if there is no plan for managing the difference between targeting the common browsers of 2015 and the common browsers of 2017.

**Comment by stefanpenner** [comment]
I think its reasonable to say, babel. these are my compile targets, and it could tell us if the build was possible given constraints.

For example.

* compile target: [chrome 47]
* feature: =>
* options: polyfill off

should be :+1: 

* compile target: [Chrome 44, IE8]
* feature: =>
* options: polyfill off

should be :-1: 

Yes this would require a list of platform capabilities, but those exist: https://kangax.github.io/compat-table/es6.

I suspect as we continue to transition into an ES2015+ world. This will be more and more valuable.

--- 

Something I have been noodling on for some time is, spec'ing out a `.platforms` file, that tooling can peer into and make the right choices.

For example:

```
// .platforms
- chrome:
  - some matrix (or ver range)
- IE:
  - 8
  - 9
  - 10
....
```

This would allow:

* CI / testem to automatically choose the required runtimes
* hinting to warn about ambient globals that do not exist in the given ENV
* compilers (like babel) to make optimal choices without the user having to know which levers to pull.

**Comment by sebmck** [comment]
Feel free to build this yourself. As I've said, you can achieve this in userland with Babel plugins. I don't see any compelling reason this should be in core as what you want is much more than what Babel provides (an entire engine/browser matrix with supported features and aliases to transformers).

**Comment by stefanpenner** [comment]
It is a recommendation, you have in the past asked me what sort of features would be of value. And as such I have added my thoughts to this one.

Babel appears to be aiming to be fairly all encompassing solution and viability of a compile target for a given feature doesn't seem out of scope.

If the issue is time required in creating the feature/platform matrix. I'm sure that can be delegates.

I am apparently not interpreting the goals correctly, please help me understand so future suggestions are more inline with the projects goals.

**Comment by sebmck** [comment]
> It is a recommendation, you have in the past asked me what sort of features would be of value. And as such I have added my thoughts to this one.

Sure, definently not saying that this isn't still the case.

> I am apparently not interpreting the goals correctly, please help me understand so future suggestions are more inline with the projects goals.

The problem is that this doesn't make sense in core other than shifting maintenance/development burden. Appropriate core feature requests are really only plugin API and feature requests now that everything is going into plugins.

What you're requesting is far too broad to be implemented in core. What you want some way for plugins to specify their compile target and then for some action to be taken on that. This is far too broad and dynamic.

As for the detection itself. Classes technically requires a bunch of object reflection methods by default, but in loose mode it doesn't. How do you even attempt to express this sort of functionality? What about chained plugins? Just because a plugin transform something to something incompatible doesn't mean that another plugin wont transform it into something that is compatible.

The actual action is very limiting too, Babel could support a set of options that then peform actions such as erroring or compiling differently depending on target. But what if you want some different behaviour? It seems like even offering anything at all is far too opinionated. You would then need to expose some way for this to be extended which is quite complex and inline with what Babel 6 is doing, everything from core would go into userland plugins anyway.

This is why the focus has gone to better APIs for userland and is why I suggested you implement this yourself so you get exactly the behaviour you want and you can be as opinionated as you want.

**Comment by sebmck** [comment]
btw if you want a faster way to iterate on and discuss then the feel free to jump into the [Babel Slack](http://slack.babeljs.io/), I'm also in the Ember Community Slack if that's useful.
